### PR TITLE
Standardized bootstrapping data.

### PR DESCRIPTION
**Pull request by ralphiee22**
## Summary
Bootstrapping data using resource layer and api calls. Handled edge cases on client side and template tags for when there is no available channels to pick from. 

## TODO
- [ ] Have tests been written for the new code?
- [ ] Has documentation been written/updated?

## Reviewer guidance
For the parameters passed into the bootstrapping method, if a call needs to be made to the server, pass in the url reversal parameters prepended with `kwargs_` (ex. `kwargs_channel_id=1`). Any query params for the url can be passed as is. (ex. `next_steps=2`)

## Documentation
To be written..
## Screenshots (if appropriate)

**Before** Accessing the learn page without bootstrapping.
screen shot 2016-10-13 at 11 50 07 am

**After** bootstrapping the recommendations.
screen shot 2016-10-13 at 11 56 22 am

### CODE REVIEW

File: kolibri/core/templatetags/kolibri_tags.py [hunk 1]
```diff
@@ -40,3 +45,46 @@ def kolibri_main_navigation():
...
+    # remove prepended string and matching items from kwargs
+    for key in kwargs.keys():
+        if kwargs_check in key:
+            item = kwargs.pop(key)
+            key = re.sub(kwargs_check, '', key)
+            reversal[key] = item
+    url = reverse('{0}-{1}'.format(base_name, route), kwargs=reversal)
+    # switch out None temporarily because invalid filtering and caching can occur
+    _replace_dict_values(None, str(''), kwargs)
+    response = client.get(url, data=kwargs)
```
Comment (indirectlylit) [inline kolibri/core/templatetags/kolibri_tags.py:96]: It seems a bit "dirty" to be using the [test HTTP client](https://docs.djangoproject.com/en/1.10/topics/testing/tools/#the-test-client) here - I don't think it was designed for this purpose.

I believe it should be possible to get a reference to the [target renderer](http://www.django-rest-framework.org/api-guide/renderers/) and [call it programmatically](http://stackoverflow.com/questions/26790981/how-to-programatically-call-a-django-rest-framework-view-within-another-view).

File: kolibri/plugins/learn/templates/learn/learn.html [hunk 2]
```diff
@@ -1,18 +1,19 @@
 {% extends "kolibri/base.html" %}
 {% load learn_tags %}
+{% load kolibri_tags %}
 
 {% block content %}
+{% kolibri_bootstrap_collection 'channel' 'ChannelResource' %}
   <script>
```
Comment (indirectlylit) [inline kolibri/plugins/learn/templates/learn/learn.html]: My intuition is that rather than having a special case for this particular model, it would be better to do it consistently.

It might actually be possible to do through the existing API - is there an API version of this lookup?

``` python
ContentNode.objects.get(parent__isnull=True)
```

If not, we could possibly add a new API call specifically for the root node.

(cc @jamalex )

File: kolibri/plugins/learn/templates/learn/learn.html [hunk 3]
```diff
@@ -1,18 +1,19 @@
...
-    var root_node_pk = String(root_node.pk);
     var model = {{kolibri}}.resources.ContentNodeResource.createModel(root_node);
     model.synced = true;
-    var collection = {{kolibri}}.resources.ContentNodeResource.createCollection({parent: root_node_pk}, JSON.parse("{{ nodes|escapejs }}"));
-    collection.synced = true;
   </script>
+  {% kolibri_bootstrap_collection 'contentnode' 'ContentNodeResource' parent=rootnode_pk kwargs_channel_id=request.COOKIES.currentChannel %}
+  {% kolibri_bootstrap_collection 'contentnode' 'ContentNodeResource' popular=user.id channel=request.COOKIES.currentChannel kwargs_channel_id=request.COOKIES.currentChannel %}
+  {% kolibri_bootstrap_collection 'contentnode' 'ContentNodeResource' resume=user.id channel=request.COOKIES.currentChannel kwargs_channel_id=request.COOKIES.currentChannel %}
+  {% kolibri_bootstrap_collection 'contentnode' 'ContentNodeResource' next_steps=user.id channel=request.COOKIES.currentChannel kwargs_channel_id=request.COOKIES.currentChannel %}
```
Comment (indirectlylit) [inline kolibri/plugins/learn/templates/learn/learn.html]: As we discussed, my recommendation would be to make each of these a unique, parameter-less tag defined either in core's `kolibri_tags.py` (e.g. `bootstrap_session`) or in a particular plugin's tag list (e.g. `bootstrap_recommended_content` in Learn).

Internally, each of these tags would be a thin wrapper around helper functions similar to `kolibri_bootstrap_collection` and `kolibri_bootstrap_model`, essentially just passing in the right parameters.

This has a few advantages:
- gives each bootstrap command a name, making it easier to scan and read
- takes complex logic out of django templates and puts it into real python files
- gets rid of the hacky `kwargs_` prefix-based interface

File: kolibri/core/templatetags/kolibri_tags.py (hunk 1 above)
Comment (jamalex) [reply kolibri/core/templatetags/kolibri_tags.py:96]: There's a lot more that happens in the API client than just calling the renderer -- running Django middleware, doing permission checks, setting up environment variables, etc (it goes through WSGI -- doing everything short of actually doing a socket-based HTTP request). All of these things are important for ensuring the data that is loaded and cached matches what would be returned by the actual API call from the client side. I'm having trouble thinking of another way to do that.

Some precedent: http://stackoverflow.com/a/4579199

File: kolibri/core/assets/src/core-actions.js [hunk 4]
```diff
@@ -350,8 +350,10 @@ function initContentSession(store, coreApp, channelId, contentId, contentKind) {
 function _setChannelState(store, coreApp, currentChannelId, channelList) {
   store.dispatch('SET_CORE_CHANNEL_LIST', channelList);
   store.dispatch('SET_CORE_CURRENT_CHANNEL', currentChannelId);
-  coreApp.resources.ContentNodeResource.setChannel(currentChannelId);
-  cookiejs.set('currentChannelId', currentChannelId);
```
Comment (ralphiee22) [inline kolibri/core/assets/src/core-actions.js:354]: Cookies should only be set if we have a valid channel id to use.

File: kolibri/core/assets/src/core-getters.js [hunk 5]
```diff
@@ -32,6 +32,10 @@ function getDefaultChannelId(channelList) {
 
 /* return the current channel object, according to vuex state */
 function getCurrentChannelObject(state) {
```
Comment (ralphiee22) [inline kolibri/core/assets/src/core-getters.js:34]: This fixes an issue on the management page (content) that would show the error box `Channel not found` since there were no available channels.

File: kolibri/plugins/learn/assets/src/actions.js [hunk 6]
```diff
@@ -218,10 +218,10 @@ function showLearnChannel(store, channelId, page = 1) {
       if (!coreGetters.getCurrentChannelObject(store.state)) {
         return;
       }
-      const nextStepsPayload = { next_steps: session.user_id, channel: channelId };
-      const popularPayload = { popular: session.user_id, channel: channelId };
-      const resumePayload = { resume: session.user_id, channel: channelId };
-      const allPayload = { kind: 'content', channel: channelId };
```
Comment (ralphiee22) [inline kolibri/plugins/learn/assets/src/actions.js:224]: Took out `channel` since it is already in the api url call.

File: kolibri/core/assets/src/core-getters.js (hunk 5 above)
Comment (indirectlylit) [reply kolibri/core/assets/src/core-getters.js:34]: I'm confused why this would return `true` when there are no channels?

Previously, the function returned either a channel object, or `undefined`. (see [`Array.find`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) docs.)

The existing logic would handle the case where `state.core.channels.list` is an empty array because calling `find` on `[]` will always return `undefined`. So it sounds like possibly the problem was that `state.core.channels.list` was undefined? If that was the case, we could do something like

```javascript
// if no channels to match against, return true
if (state.core.channels.list === undefined) {
  return undefined;
}
```

But that would still return the opposite of what's here - `undefined` instead of `true`...

File: kolibri/core/assets/src/core-actions.js [hunk 7]
```diff
@@ -350,8 +350,10 @@ function initContentSession(store, coreApp, channelId, contentId, contentKind) {
...
+  if (currentChannelId !== null) {
+    coreApp.resources.ContentNodeResource.setChannel(currentChannelId);
+    cookiejs.set('currentChannelId', currentChannelId);
+  }
```
Comment (indirectlylit) [inline kolibri/core/assets/src/core-actions.js]: If

```javascript
currentChannelId === null
```

we might want to actually

```javascript
cookiejs.remove('currentChannelId');
```

(If the user previously had a cookie set, we probably don't want that to be used in the future).

File: kolibri/core/assets/src/core-getters.js (hunk 5 above)
Comment (ralphiee22) [reply kolibri/core/assets/src/core-getters.js:34]: My main thing I was trying to do was not get the error box here for having no channels, say in the first initial load of kolibri app.
https://github.com/learningequality/kolibri/blob/develop/kolibri/core/assets/src/core-actions.js#L355
`undefined` maps to being false, I presume?
Maybe the logic for handling when there are no channels vs. when we can not find the queried channel should be different in this case.

**Comment by indirectlylit** [comment]
PR submitted: https://github.com/ralphiee22/kolibri/pull/5

**Comment by indirectlylit** [comment]
🎉
